File: examples/ios-download/AntDownload/AntNode.swift
import Foundation
import SwiftUI

/// Swift wrapper around the hand-written C API in `crates/ant-ffi/`.
///
/// Holds a single `AntHandle*` for the lifetime of the app. `start()`
/// spins up the embedded node on a background task; `download(...)`
/// blocks on a cooperative Swift task while the Rust side runs the
/// retrieval on its own Tokio runtime. The FFI is thread-safe, so
/// multiple concurrent `download` calls share the node's chunk cache
/// and retrieval pipeline.
///
/// `AntHandle` is an opaque forward-declared struct in C
/// (`typedef struct AntHandle AntHandle;`) — Swift imports that as
/// `OpaquePointer`, so that is what we carry around here.
@MainActor
final class AntNode: ObservableObject {
    enum Status: Equatable {
        case idle
        case starting
        case ready
        case failed(String)

        var label: String {
            switch self {
            case .idle: return "Idle"
            case .starting: return "Starting…"
            case .ready: return "Ready"
            case .failed(let msg): return "Failed: \(msg)"
            }
        }

        var color: Color {
            switch self {
            case .idle: return .secondary
            case .starting: return .orange
            case .ready: return .green
            case .failed: return .red
            }
        }

        var isReady: Bool {
            if case .ready = self { return true }
            return false
        }
    }

    @Published private(set) var status: Status = .idle
    @Published private(set) var peerCount: Int = 0
    /// Agent string reported by the embedded node's status snapshot,
    /// e.g. `ant-ffi/0.4.0`. Sourced from `ant_agent_string`, set
    /// once the FFI handle is up. Used in the Player header so the
    /// version we display reflects the *running* binary rather than
    /// the iOS bundle's `CFBundleShortVersionString`.
    @Published private(set) var agentString: String = ""
    /// Live progress view of the in-flight `download(reference:)` call,
    /// updated ~4×/s from a poll loop. `nil` whenever no download is
    /// running (or the last one finished and its state was cleared).
    @Published private(set) var downloadProgress: DownloadProgress? = nil

    private var handle: OpaquePointer?
    private var peerPollTask: Task<Void, Never>?
    private var progressPollTask: Task<Void, Never>?

    func start() async {
        guard case .idle = status else { return }
        status = .starting

        let dataDir = Self.resolveDataDir()
        Self.seedPeerstoreIfNeeded(in: dataDir)
        let path = dataDir.path

        let result: InitOutcome = await Task.detached(priority: .userInitiated) {
            var errPtr: UnsafeMutablePointer<CChar>? = nil
            let raw: OpaquePointer? = path.withCString { cpath in
                ant_init(cpath, &errPtr)
            }
            if let raw = raw {
                return .ready(raw)
            }
            let msg = errPtr.flatMap { String(cString: $0) } ?? "unknown error"
            if let errPtr = errPtr {
                ant_free_string(errPtr)
            }
            return .failed(msg)
        }.value

        switch result {
        case .ready(let h):
            handle = h
            status = .ready
            agentString = Self.readAgentString(handle: h)
            startPollingPeers(handle: h)
        case .failed(let msg):
            status = .failed(msg)
        }
    }

    /// One-shot pull of the running node's `agent` field. The status
    /// snapshot's agent string is fixed at `ant_init` time, so polling
    /// would burn cycles for no signal.
    private static func readAgentString(handle: OpaquePointer) -> String {
        guard let raw = ant_agent_string(handle) else { return "" }
        defer { ant_free_string(raw) }
        return String(cString: raw)
    }

    func shutdown() async {
        guard let h = handle else { return }
        handle = nil
        peerPollTask?.cancel()
        peerPollTask = nil
        progressPollTask?.cancel()
        progressPollTask = nil
        peerCount = 0
        agentString = ""
        downloadProgress = nil
        await Task.detached(priority: .userInitiated) {
            ant_shutdown(h)
        }.value
        status = .idle
    }

    /// Poll `ant_peer_count` 5×/s so the UI row tracks the BZZ
    /// handshake progress. `watch::Receiver` is not friendly to a
    /// C-ABI boundary, so polling is the simplest stable shape — and
    /// cheap: each call is a `RwLock` read of the shared snapshot.
    private func startPollingPeers(handle: OpaquePointer) {
        peerPollTask?.cancel()
        peerPollTask = Task { [weak self] in
            while !Task.isCancelled {
                let count = Int(ant_peer_count(handle))
                await MainActor.run {
                    guard let self = self, self.handle == handle else { return }
                    self.peerCount = max(0, count)
                }
                try? await Task.sleep(nanoseconds: 200_000_000)
            }
        }
    }

    /// Abort the in-flight `download(reference:)`. No-op when no
    /// download is running. The racing `download` call throws
    /// `AntError.download("download canceled")` so the view can
    /// surface it (or swallow it, if it already knows the user hit
    /// Pause or Cancel).
    ///
    /// Caveat: this only stops our Swift/FFI wait loop. The embedded
    /// node's `run_get_bzz` task is not cross-request cancellable
    /// today — it keeps fetching remaining chunks in the background
    /// and populates the in-memory chunk cache. That's what makes
    /// "Pause → Download" feel like a resume. A full cancellation
    /// token through `ControlCommand` is a follow-up (PLAN §11).
    func cancelDownload() {
        guard let h = handle else { return }
        ant_cancel_download(h)
    }

    /// Bulk download without touching the published `downloadProgress`
    /// slot — used for ancillary fetches (album art, manifest probes)
    /// where surfacing a HUD over the main playback UI would be
    /// noisy. The single-flight progress poller is reserved for the
    /// user-visible `download(reference:)` path.
    func downloadQuiet(reference: String) async throws -> Data {
        guard let h = handle else {
            throw AntError.notReady
        }
        return try await Task.detached(priority: .utility) {
            var len: Int = 0
            var errPtr: UnsafeMutablePointer<CChar>? = nil
            let body = reference.withCString { cref in
                ant_download(h, cref, &len, &errPtr)
            }
            if let body = body {
                let data = Data(bytes: body, count: len)
                ant_free_buffer(body, len)
                return data
            }
            let msg = errPtr.flatMap { String(cString: $0) } ?? "unknown download error"
            if let errPtr = errPtr {
                ant_free_string(errPtr)
            }
            throw AntError.download(msg)
        }.value
    }

    func download(reference: String) async throws -> Data {
        guard let h = handle else {
            throw AntError.notReady
        }
        startPollingProgress(handle: h)
        defer {
            progressPollTask?.cancel()
            progressPollTask = nil
            // Leave the final sample visible for ~400 ms so the UI
            // can freeze on "100 %" instead of snapping back to nothing
            // the instant ant_download returns.
            Task { [weak self] in
                try? await Task.sleep(nanoseconds: 400_000_000)
                await MainActor.run {
                    guard let self = self, self.progressPollTask == nil else { return }
                    self.downloadProgress = nil
                }
            }
        }
        return try await Task.detached(priority: .userInitiated) {
            var len: Int = 0
            var errPtr: UnsafeMutablePointer<CChar>? = nil
            let body = reference.withCString { cref in
                ant_download(h, cref, &len, &errPtr)
            }
            if let body = body {
                // Copy into Swift-managed storage so the caller can
                // drop the Rust-owned buffer immediately.
                let data = Data(bytes: body, count: len)
                ant_free_buffer(body, len)
                return data
            }
            let msg = errPtr.flatMap { String(cString: $0) } ?? "unknown download error"
            if let errPtr = errPtr {
                ant_free_string(errPtr)
            }
            throw AntError.download(msg)
        }.value
    }

    /// Start a background poll that reads `ant_download_progress` 4×/s,
    /// derives a smoothed throughput, and publishes the result on the
    /// main actor. Cancelled in `download(reference:)`'s defer.
    private func startPollingProgress(handle: OpaquePointer) {
        progressPollTask?.cancel()
        // Start with a fresh zero-progress sample so the UI can render
        // the HUD immediately (chunks_done=0, total=0) rather than
        // waiting for the first tick — otherwise the first ~150 ms
        // after hitting Download shows no feedback at all.
        downloadProgress = DownloadProgress(
            bytesDone: 0,
            totalBytes: 0,
            chunksDone: 0,
            totalChunks: 0,
            elapsedMs: 0,
            peersUsed: 0,
            inFlight: 0,
            cacheHits: 0,
            throughputBytesPerSec: 0
        )
        progressPollTask = Task { [weak self] in
            // Sliding-window mean over the last `windowSeconds` of
            // (timestamp, bytes_done) samples. Compared to the EMA we
            // had before (alpha = 0.3 → ~830 ms time constant), a
            // straight 3 s window is a) much calmer when chunks land
            // in `buffer_unordered` bursts and b) easier to reason
            // about: the readout literally is "average MiB/s across
            // the last 3 s", which is what users expect a download
            // speedometer to mean. The chunks / peers / in-flight
            // counters keep updating at the full 250 ms cadence
            // because the window only smooths the rate, not the rest.
            let windowSeconds: TimeInterval = 3.0
            var samples: [(at: Date, bytes: UInt64)] = []
            var progress = AntProgress()
            while !Task.isCancelled {
                let rc = ant_download_progress(handle, &progress)
                if rc == 0 {
                    let now = Date()
                    samples.append((now, progress.bytes_done))
                    let cutoff = now.addingTimeInterval(-windowSeconds)
                    if let firstFresh = samples.firstIndex(where: { $0.at >= cutoff }),
                       firstFresh > 1 {
                        // Keep one stale sample so the window has a
                        // baseline at the trailing edge — without it,
                        // dropping every "too old" entry would leave
                        // us computing rate over only the latest few
                        // hundred ms again.
                        samples.removeFirst(firstFresh - 1)
                    }
                    let bytesPerSec: Double
                    if let oldest = samples.first, samples.count >= 2 {
                        let dt = now.timeIntervalSince(oldest.at)
                        if dt > 0.25 {
                            let delta = progress.bytes_done >= oldest.bytes
                                ? Double(progress.bytes_done - oldest.bytes)
                                : 0
                            bytesPerSec = delta / dt
                        } else {
                            bytesPerSec = 0
                        }
                    } else {
                        bytesPerSec = 0
                    }
                    let snapshot = DownloadProgress(
                        bytesDone: progress.bytes_done,
                        totalBytes: progress.total_bytes,
                        chunksDone: progress.chunks_done,
                        totalChunks: progress.total_chunks,
                        elapsedMs: progress.elapsed_ms,
                        peersUsed: progress.peers_used,
                        inFlight: progress.in_flight,
                        cacheHits: progress.cache_hits,
                        throughputBytesPerSec: bytesPerSec
                    )
                    await MainActor.run {
                        guard let self = self, self.handle == handle else { return }
                        self.downloadProgress = snapshot
                    }
                }
                try? await Task.sleep(nanoseconds: 250_000_000)
            }
        }
    }

    /// Synchronously list a `bzz://` manifest. Returns the decoded
    /// `Manifest` on success; throws on FFI / decode failure. Runs on a
    /// detached task so the caller's MainActor isn't blocked while the
    /// node walks the mantaray tree.
    func listManifest(reference: String) async throws -> Manifest {
        guard let h = handle else { throw AntError.notReady }
        return try await Task.detached(priority: .userInitiated) {
            var errPtr: UnsafeMutablePointer<CChar>? = nil
            let raw = reference.withCString { cref in
                ant_list_bzz(h, cref, &errPtr)
            }
            if let raw = raw {
                let json = String(cString: raw)
                ant_free_string(raw)
                guard let data = json.data(using: .utf8) else {
                    throw AntError.download("manifest json was not utf-8")
                }
                return try JSONDecoder().decode(Manifest.self, from: data)
            }
            let msg = errPtr.flatMap { String(cString: $0) } ?? "unknown manifest error"
            if let errPtr = errPtr { ant_free_string(errPtr) }
            throw AntError.download(msg)
        }.value
    }

    /// Open a streaming session for one file in a manifest. The session
    /// frees its underlying `AntStream*` in `deinit`; do not let the
    /// node be torn down while a session is alive (callers hold the
    /// session on `PlayerEngine`, which is bound to the same lifetime
    /// as `AntNode`).
    func openStream(reference: String) async throws -> AntStreamSession {
        guard let h = handle else { throw AntError.notReady }
        let session: AntStreamSession? = await Task.detached(priority: .userInitiated) {
            AntStreamSession(handle: h, bzzReference: reference)
        }.value
        guard let session = session else {
            throw AntError.download("failed to open stream for \(reference)")
        }
        return session
    }

    /// Expose the raw handle so the resource loader can call FFI
    /// directly without bouncing through MainActor for every byte.
    /// Nil whenever the node is not `.ready`.
    var rawHandle: OpaquePointer? { handle }

    private static func resolveDataDir() -> URL {
        let fm = FileManager.default
        let base = (try? fm.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )) ?? fm.temporaryDirectory
        let dir = base.appendingPathComponent("ant", isDirectory: true)
        try? fm.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    /// Cold-start bootstrap for the embedded node. Historically
    /// (ant 0.4.x) mainnet bootnodes that had moved to bee 2.8
    /// rejected our outbound BZZ handshake — we only spoke the
    /// `14.0.0` protocol id while bee 2.8 only listens on `15.0.0`
    /// — so a fresh install with no `peers.json` never left
    /// `peer_set_size = 0`. Ant 0.5.0+ speaks both wire versions
    /// (defaults to V15, falls back to V14), but we keep seeding
    /// `peers.json` here so the pipeline can dial hive hints in
    /// parallel with the first round of cold-start handshakes.
    private static func seedPeerstoreIfNeeded(in dataDir: URL) {
        let dest = dataDir.appendingPathComponent("peers.json")
        if FileManager.default.fileExists(atPath: dest.path),
           let data = try? Data(contentsOf: dest),
           let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let peers = obj["peers"] as? [Any],
           !peers.isEmpty {
            return
        }
        guard let bundled = Bundle.main.url(forResource: "peers.seed", withExtension: "json") else {
            NSLog("[ant] peers.seed.json missing from app bundle")
            return
        }
        do {
            if FileManager.default.fileExists(atPath: dest.path) {
                try FileManager.default.removeItem(at: dest)
            }
            try FileManager.default.copyItem(at: bundled, to: dest)
        } catch {
            NSLog("[ant] failed to seed peers.json: %@", "\(error)")
        }
    }
}

private enum InitOutcome {
    case ready(OpaquePointer)
    case failed(String)
}

/// Swift-friendly value type mirroring `AntProgress` plus a smoothed
/// throughput field derived on the Swift side (the Rust FFI doesn't
/// compute rates — it just ships the raw counters).
struct DownloadProgress: Equatable {
    var bytesDone: UInt64
    var totalBytes: UInt64
    var chunksDone: UInt64
    var totalChunks: UInt64
    var elapsedMs: UInt64
    var peersUsed: UInt32
    var inFlight: UInt32
    var cacheHits: UInt64
    /// Exponential-moving-average of `delta_bytes / delta_time` over
    /// the poll samples. `0` until we have at least two samples.
    var throughputBytesPerSec: Double

    /// `0.0 ... 1.0` progress fraction, or `nil` while the data root
    /// hasn't landed yet (which is when `totalBytes == 0`).
    var fraction: Double? {
        guard totalBytes > 0 else { return nil }
        return min(1.0, Double(bytesDone) / Double(totalBytes))
    }
}

enum AntError: LocalizedError {
    case notReady
    case download(String)

    var errorDescription: String? {
        switch self {
        case .notReady: return "Node is not ready"
        case .download(let msg): return msg
        }
    }
}

File: examples/android-download/app/src/main/java/at/vibing/ant/downloadsmoke/AntNode.kt
package at.vibing.ant.downloadsmoke

import android.content.Context
import android.os.SystemClock
import androidx.compose.runtime.Immutable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

/**
 * Kotlin wrapper around the JNI surface in `crates/ant-ffi/src/jni.rs`.
 *
 * Sibling to `AntNode.swift` in the iOS smoke test. Holds a single
 * `nativeHandle: Long` for the lifetime of the app, exposes
 * `StateFlow`-backed observables for the status / peer count /
 * agent string / progress fields the UI subscribes to, and runs every
 * blocking JNI call on `Dispatchers.IO` so the main thread is never
 * caught violating Android's `NetworkOnMainThreadException` policy.
 *
 * Lifecycle:
 *
 *  - Construct a single instance in the [AntDownloadApp.onCreate] hook
 *    (or the Application equivalent). Construction is cheap — it only
 *    loads `libant_ffi.so`.
 *  - Call [start] once Compose has the `Context` to hand us a writable
 *    data dir. Subsequent calls are a no-op.
 *  - Call [download] for each retrieval; concurrent calls share the
 *    embedded node's chunk cache and command channel.
 *  - Call [shutdown] when the process is leaving for good (rare in
 *    Android — usually the OS just kills the process).
 *
 * Threading:
 *
 *  - All `external fun` calls are blocking and live on `Dispatchers.IO`.
 *  - StateFlow updates flip back to the main dispatcher implicitly via
 *    `MutableStateFlow.value =`.
 *  - The handle is plain `Long` (`*mut AntHandle` on the Rust side)
 *    and accessed without locking; the Rust state machine is internally
 *    `Send + Sync`.
 */
class AntNode {

    /**
     * Discriminated status for the embedded node. Mirrors the iOS
     * `AntNode.Status` enum — same names, same colours, same legible
     * `label` strings so logcat / Xcode console lines read identically.
     */
    @Immutable
    sealed interface Status {
        val label: String

        data object Idle : Status { override val label: String = "Idle" }
        data object Starting : Status { override val label: String = "Starting…" }
        data object Ready : Status { override val label: String = "Ready" }
        data class Failed(val message: String) : Status {
            override val label: String get() = "Failed: $message"
        }

        val isReady: Boolean get() = this is Ready
    }

    private val _status = MutableStateFlow<Status>(Status.Idle)
    val status: StateFlow<Status> = _status.asStateFlow()

    private val _peerCount = MutableStateFlow(0)
    val peerCount: StateFlow<Int> = _peerCount.asStateFlow()

    private val _agentString = MutableStateFlow("")
    val agentString: StateFlow<String> = _agentString.asStateFlow()

    /**
     * Live progress view of the in-flight `download(reference)` call,
     * updated ~4×/s from a poll job. `null` whenever no download is
     * running (or the last one finished and its state was cleared).
     * Direct mirror of the iOS `@Published downloadProgress`.
     */
    private val _downloadProgress = MutableStateFlow<DownloadProgress?>(null)
    val downloadProgress: StateFlow<DownloadProgress?> = _downloadProgress.asStateFlow()

    /**
     * Raw `*mut AntHandle` cast to `jlong`, populated by [start]'s
     * call to [nativeInit]. Zero whenever the node is not started.
     * Volatile so the polling job sees the post-`shutdown` zero
     * without holding a lock — same threading discipline as the iOS
     * `private var handle: OpaquePointer?` field.
     */
    @Volatile
    private var nativeHandle: Long = 0L

    /**
     * Coroutine scope rooted on a SupervisorJob so a cancelled poll
     * task doesn't take down the rest. Cancelled in [shutdown].
     */
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    /**
     * Background poller that reads `nativePeerCount` 5×/s once the
     * node is `Ready`. Direct mirror of the iOS `peerPollTask`.
     */
    private var peerPollJob: Job? = null

    /**
     * Background poller that reads `nativeDownloadProgress` 4×/s while
     * a download is in flight. Started in [download], cancelled in
     * its `finally` block. Mirror of iOS `progressPollTask`.
     */
    private var progressPollJob: Job? = null

    /**
     * Spin up the embedded node. Idempotent: subsequent calls while
     * status != Idle are no-ops, matching iOS `AntNode.start()`.
     *
     * @param context any Context — used solely to look up
     *   `filesDir.absolutePath`, which we hand the FFI as the
     *   data-dir for `identity.json` / `peers.json` persistence.
     */
    suspend fun start(context: Context) {
        if (_status.value !is Status.Idle) return
        _status.value = Status.Starting

        val dataDir: String = withContext(Dispatchers.IO) {
            // Application Support equivalent on Android. Files here
            // survive app upgrades and aren't user-visible. Sized
            // about right for a peer snapshot + identity (~2 KB).
            val dir = File(context.filesDir, "ant")
            dir.mkdirs()
            dir.absolutePath
        }

        val outcome: InitOutcome = withContext(Dispatchers.IO) {
            try {
                val handle = nativeInit(dataDir)
                if (handle == 0L) {
                    InitOutcome.Failed("nativeInit returned 0 without throwing")
                } else {
                    InitOutcome.Ready(handle)
                }
            } catch (e: AntException) {
                InitOutcome.Failed(e.message ?: "unknown error")
            } catch (t: Throwable) {
                InitOutcome.Failed("unexpected: ${t.message ?: t.javaClass.simpleName}")
            }
        }

        when (outcome) {
            is InitOutcome.Ready -> {
                nativeHandle = outcome.handle
                _status.value = Status.Ready
                _agentString.value = readAgent(outcome.handle)
                startPollingPeers()
            }
            is InitOutcome.Failed -> {
                _status.value = Status.Failed(outcome.message)
            }
        }
    }

    /** Tear down the embedded node. Safe to call when not started. */
    suspend fun shutdown() {
        val h = nativeHandle
        if (h == 0L) return
        nativeHandle = 0L
        peerPollJob?.cancel()
        peerPollJob = null
        progressPollJob?.cancel()
        progressPollJob = null
        _peerCount.value = 0
        _agentString.value = ""
        _downloadProgress.value = null
        withContext(Dispatchers.IO) { nativeShutdown(h) }
        _status.value = Status.Idle
    }

    /**
     * Block (on `Dispatchers.IO`) until the requested reference is
     * fully retrieved. Returns the raw bytes; the caller is
     * responsible for whatever rendering / persistence they want
     * to do next. Throws [AntException] on any FFI-reported error,
     * mirroring iOS `AntError.download(String)`.
     *
     * While the call is in flight, [downloadProgress] is populated
     * by a 4 Hz poll job. The slot is cleared back to `null` ~400 ms
     * after `download` returns so the UI can hold the final "100 %"
     * frame instead of snapping back to nothing the instant the FFI
     * call resolves — same hold-time the iOS smoke uses.
     */
    suspend fun download(reference: String): ByteArray {
        val h = nativeHandle
        if (h == 0L) throw AntException("Node is not ready")
        startPollingProgress(h)
        try {
            return withContext(Dispatchers.IO) {
                nativeDownload(h, reference)
            }
        } finally {
            progressPollJob?.cancel()
            progressPollJob = null
            // Hold the final progress sample on screen briefly so a
            // sub-second download (cache hit) doesn't flash a blank
            // card before the result lands. Mirrors the 400 ms hold
            // in `AntNode.swift` (download(reference:)).
            scope.launch {
                delay(400L)
                if (progressPollJob == null) {
                    _downloadProgress.value = null
                }
            }
        }
    }

    /**
     * Request cancellation of the in-flight `download(reference)`. No-op
     * when no download is running. The racing `download` call throws
     * `AntException("download canceled")` on its next ack-loop
     * iteration (≤ ~150 ms under healthy retrieval).
     *
     * Caveat: this only stops our Kotlin/FFI wait loop. The embedded
     * node's `run_get_bzz` task is not cross-request cancellable
     * today — it keeps fetching remaining chunks in the background
     * and populates the in-memory chunk cache. That's what makes a
     * subsequent retry of the same reference feel like a resume.
     * Same caveat as the iOS smoke (`AntNode.cancelDownload()`).
     */
    fun cancelDownload() {
        val h = nativeHandle
        if (h == 0L) return
        nativeCancelDownload(h)
    }

    /**
     * One-shot pull of the running node's `agent` field. The status
     * snapshot fixes this at `nativeInit` time, so polling would
     * burn cycles for no signal. Same shape as iOS
     * `AntNode.readAgentString(handle:)`.
     */
    private fun readAgent(handle: Long): String {
        return try {
            nativeAgentString(handle).orEmpty()
        } catch (t: Throwable) {
            ""
        }
    }

    /**
     * 5 Hz peer-count poll. Cheap (a `RwLock` read of the shared
     * status snapshot inside the embedded node), so the polling
     * cost on a low-end phone is dominated by the JNI transition,
     * not the read itself.
     */
    private fun startPollingPeers() {
        peerPollJob?.cancel()
        peerPollJob = scope.launch {
            while (true) {
                val h = nativeHandle
                if (h == 0L) break
                val count = try {
                    nativePeerCount(h)
                } catch (t: Throwable) {
                    -1
                }
                _peerCount.value = if (count < 0) 0 else count
                delay(200L)
            }
        }
    }

    /**
     * Read `nativeDownloadProgress(handle)` 4×/s, smooth the
     * throughput over a 3-second sliding window, and publish the
     * result on `_downloadProgress`. Direct port of the iOS
     * `AntNode.startPollingProgress(handle:)` loop in
     * `examples/ios-download/AntDownload/AntNode.swift`.
     */
    private fun startPollingProgress(handle: Long) {
        progressPollJob?.cancel()
        // Seed a fresh zero-progress sample so the UI can render the
        // HUD on the *next* recompose rather than waiting up to one
        // poll interval — same pattern as the iOS code.
        _downloadProgress.value = DownloadProgress.empty()
        progressPollJob = scope.launch {
            // 3 s sliding window over (timestamp_ms, bytes_done) pairs.
            // Compared to an EMA, a straight window is much calmer
            // when chunks land in `buffer_unordered` bursts, and the
            // readout literally is "average MiB/s over the last 3 s",
            // which is what users expect a download speedometer to
            // mean. The chunks / peers / in-flight counters keep
            // updating at the full 250 ms cadence because the window
            // only smooths the rate, not the rest.
            val windowMs = 3_000L
            val samples = ArrayDeque<Sample>()
            while (true) {
                val h = nativeHandle
                if (h == 0L) break
                val raw: LongArray? = try {
                    nativeDownloadProgress(h)
                } catch (t: Throwable) {
                    null
                }
                if (raw != null && raw.size >= 9) {
                    val now = SystemClock.elapsedRealtime()
                    val bytesDone = raw[1]
                    samples.addLast(Sample(at = now, bytes = bytesDone))
                    val cutoff = now - windowMs
                    while (samples.size > 1 && samples.first().at < cutoff) {
                        samples.removeFirst()
                    }
                    val bytesPerSec: Double = if (samples.size >= 2) {
                        val oldest = samples.first()
                        val dtMs = (now - oldest.at).coerceAtLeast(0L)
                        if (dtMs >= 250L) {
                            val deltaBytes = (bytesDone - oldest.bytes).coerceAtLeast(0L)
                            deltaBytes.toDouble() * 1_000.0 / dtMs.toDouble()
                        } else 0.0
                    } else 0.0
                    _downloadProgress.value = DownloadProgress(
                        inProgress = raw[0] != 0L,
                        bytesDone = raw[1],
                        totalBytes = raw[2],
                        chunksDone = raw[3],
                        totalChunks = raw[4],
                        elapsedMs = raw[5],
                        peersUsed = raw[6].toInt(),
                        inFlight = raw[7].toInt(),
                        cacheHits = raw[8],
                        throughputBytesPerSec = bytesPerSec,
                    )
                }
                delay(250L)
            }
        }
    }

    private sealed interface InitOutcome {
        data class Ready(val handle: Long) : InitOutcome
        data class Failed(val message: String) : InitOutcome
    }

    private data class Sample(val at: Long, val bytes: Long)

    // -----------------------------------------------------------------
    // JNI surface — mirror of `crates/ant-ffi/src/jni.rs`.
    //
    // Each `external fun` resolves to a `Java_at_vibing_ant_downloadsmoke_AntNode_<name>`
    // symbol exported from `libant_ffi.so` (which lives in
    // `app/src/main/jniLibs/<abi>/`, populated by
    // `cargo xtask build-android-*`). The companion object's static
    // initializer loads the library exactly once per process — the
    // platform-default behaviour of `System.loadLibrary` matches the
    // iOS staticlib model where all symbols are baked into the app
    // binary by the time the first activity is created.
    //
    // `external` methods are static (`@JvmStatic` via companion
    // object) so the JNI symbol mangling lines up with the Rust
    // exports, which are written as `Java_…_AntNode_native…` not
    // `…_AntNode_00024Companion_native…` (the latter would be the
    // mangling for non-static functions on `AntNode.Companion`).
    // -----------------------------------------------------------------

    companion object {
        init {
            // Triggers `dlopen("libant_ffi.so")` against the right
            // ABI subdirectory. Idempotent within a process — calling
            // `loadLibrary` twice with the same name is a documented
            // no-op since Android 4.x.
            System.loadLibrary("ant_ffi")
        }

        @JvmStatic external fun nativeInit(dataDir: String): Long
        @JvmStatic external fun nativeDownload(handle: Long, reference: String): ByteArray
        @JvmStatic external fun nativePeerCount(handle: Long): Int
        @JvmStatic external fun nativeAgentString(handle: Long): String?
        /**
         * Returns a 9-element snapshot of the in-flight download's
         * progress counters, or `null` on a null handle. See the
         * Rust side (`crates/ant-ffi/src/jni.rs`) for the exact
         * field order — the Kotlin layer in [startPollingProgress]
         * reconstructs a [DownloadProgress] from this array.
         */
        @JvmStatic external fun nativeDownloadProgress(handle: Long): LongArray?
        /** Returns 0 on success, -1 on a null handle. */
        @JvmStatic external fun nativeCancelDownload(handle: Long): Int
        @JvmStatic external fun nativeShutdown(handle: Long)
    }
}
